### ISSUE TITLE
Semantic Versioning needs to be comprehensive

### ISSUE THREAD

**Issue by fooman** [magento/magento2#4228]
I believe Semantic Versioning is meant to deliver the M2 product goal of easier upgrades.

And currently the only definition of what Magento considers to be versioned semantically is, according to [this](http://devdocs.magento.com/guides/v2.0/architecture/versioning.html), anything annotated with `@api`. 

Unfortunately the `@api` coverage is not comprehensive. Case in point, as of 2.0.x, it is currently not possible to write a single extension which is 100% `@api` compatible starting with `\Magento\Framework\Component\ComponentRegistrar::register()` and `urn:magento:framework:Module/etc/module.xsd`.

Since extensions are being written for Magento 2 and merchants are launching on M2 this in my opinion leaves two options going forward, acknowledging that there is still some way to go until `@api` becomes comprehensive:

1.) leave Semantic Versioning policy as is - limited to `@api`
2.) extend Semantic Versioning policy to cover all released code (this is what I am advocating for)

Option 1.) will again lead to two outcomes:

1.) developers will need to depend on specific versions down to the patch level of Magento core modules, which delivers zero value of doing semantic versioning in the first place and lots of developer resources are needed for every single patch upgrade
2.) stores will break because the semantic versioning limitations were misunderstood by developers and non-sanctioned methods/classes were used

Option 2.) I think will have the following outcomes:

1.) Magento will release more breaking changes as per the version numbering while refactoring to achieve comprehensive `@api` coverage. However being told about a breaking change is better than having to check every single patch release for a possible breakage outside of `@api`. In addition the developer experience could be vastly improved by keeping a change log on the magento core module level indicating what breaking changes happened. This change log can be more comprehensive than the one at the Magento product level
2.) Store upgrades are safer because stores won't be able to upgrade until all extensions have dealt with breaking changes including those not covered by `@api`

In summary:

Keeping Semantic Versioning policy as is: bad for developers and bad for merchants
Comprehensive coverage of Semantic Versioning: good for merchants, okay for developers

**Comment by Vinai** [comment]
Thanks @fooman!
In order to get comprehensive `@api` coverage breaking changes are required, for example to add new methods to existing interfaces.  
Once comprehensive coverage is achieved, the SemVer policy could be changed.

**Comment by maksek** [comment]
Thanks @fooman. Comprehensive coverage of Semantic Versioning may work in full throttle with complete module decoupling. In M2 we still have pretty wide dependencies between modules, and making major version change in one module, requires to update dependency version in another. Also Magento 2 has own versioning strategy 2.0.x, 2.1.x. in addition to each module version, and making patch for product, may have minor change in module. Did you try to project how it will work in such cases? Any feedback on this?

**Comment by fooman** [comment]
As I understood it so far the product version number for Magento 2.x is only for Marketing purposes and can be ignored when talking about Semantic Versioning. However this comment here 

> Also Magento 2 has own versioning strategy 2.0.x, 2.1.x. in addition to each module version, and making patch for product, may have minor change in module. 

seems to indicate that 2.0.x and 2.1.x patch releases are meant to include only patch level releases from your side.

I don' think you will be able to have it both ways. Either the Magento product version number is separate or it is not. For example what about a fix that you urgently need to push out that is breaking? If you followed semantic versioning on the Magento project level you would need to release 3.0. 

Taken the above together I think 2.0.x, 2.1.x releases need to be allowed to include modules with minor and breaking changes. The ideal scenario would of course be that the 2.0.x / 2.1.x branches only get patch level fixes for modules but this would not always be doable as per the above example.

> In M2 we still have pretty wide dependencies between modules, and making major version change in one module, requires to update dependency version in another.

yeap, however if the definition of semantic versioning is comprehensive Magento itself can trust that a constraint like ~100.0.2 or 100.0.\* will not break things making your life easier as well by not having to update all other dependencies just because you updated one module. 

At the start of a release candidate period you could define the exact module version number mix to provide certainty what features will be released.

**Comment by antonkril** [comment]
The plan was to release small set of @api annotated classes and later annotate classes requested by community. We did not get many requests for marking with @api, but we will be marking more classes in the next release.

If we apply SemVer to all code, we will be releasing lots of major an minor version bumps. It will force many extension developers to recheck their extensions even if they follow our development guidelines. 

So we plan to keep SemVer limited to @api but increase @api coverage. 

We're also discussing idea to analyze extensions for usage of Core Apis, and mark most popular apis with @api

**Comment by fooman** [comment]
Thanks @antonkril for the reply. 

> The plan was to release small set of @api annotated classes and later annotate classes requested by community. 

I am not sure this plan was well understood by the community. Also the big complication that I see in all this is that what I have heard from Magento is that only the Magento_Customer module is considered "done" ie fully refactored. This leaves 101 modules on the todo list. Until it is clear what the plan for these are, it does not make sense to add `@api` annotations to these in my opinion.

> If we apply SemVer to all code, we will be releasing lots of major an minor version bumps. It will force many extension developers to recheck their extensions 

Correct - that is preferable since if you don't SemVer all the code, extensions need to get rechecked on every release anyway

> even if they follow our development guidelines

I am fairly confident that there are currently 0 extensions in existence that are able to function when only using sanctioned classes/methods (read `@api` annotated) - see `\Magento\Framework\Component\ComponentRegistrar::register()`

Let's make this concrete:

### version 2.0.0

```
phploc app/code/Magento/
phploc 3.0.1 by Sebastian Bergmann.

Directories                                       3759
Files                                             8752


Structure
  Namespaces                                      3631
  Interfaces                                       531
  Traits                                             0
  Classes                                         8042
    Abstract Classes                               317 (3.94%)
    Concrete Classes                              7725 (96.06%)
  Methods                                        41629
    Scope
      Non-Static Methods                         41565 (99.85%)
      Static Methods                                64 (0.15%)
    Visibility
      Public Methods                             33334 (80.07%)
      Non-Public Methods                          8295 (19.93%)
  Functions                                        269
    Named Functions                                  4 (1.49%)
    Anonymous Functions                            265 (98.51%)
  Constants                                       3764
    Global Constants                                 0 (0.00%)
    Class Constants                               3764 (100.00%)
```

`grep -r "@api" app/code/Magento/* | tee >(wc -l)`
723

`grep -r "namespace .*\\Api" app/code/Magento/* | tee >(wc -l)`
316

So even assuming the @api covers complete classes instead of methods and the above hits do not overlap you have a coverage of 13% (1039/7725).

### **2.0.6**

```
phploc app/code/Magento/
phploc 3.0.1 by Sebastian Bergmann.

Directories                                       3764
Files                                             8777

Structure
  Namespaces                                      3635
  Interfaces                                       537
  Traits                                             0
  Classes                                         8060
    Abstract Classes                               317 (3.93%)
    Concrete Classes                              7743 (96.07%)
  Methods                                        41838
    Scope
      Non-Static Methods                         41774 (99.85%)
      Static Methods                                64 (0.15%)
    Visibility
      Public Methods                             33507 (80.09%)
      Non-Public Methods                          8331 (19.91%)
```

`grep -r "@api" app/code/Magento/* | tee >(wc -l)`
723

`grep -r "namespace .*\\Api" app/code/Magento/* | tee >(wc -l)`
319

which also gives 2.0.6 roughly 13% - this is far away from Semantic Versioning 

> Software using Semantic Versioning MUST declare a public API. This API could be declared in the code itself or exist strictly in documentation. However it is done, it should be precise and **comprehensive.**

**Comment by fooman** [comment]
Collecting some random examples here to show that this is not theoretical

> stores will break because the semantic versioning limitations were misunderstood by developers and non-sanctioned methods/classes were used 

https://twitter.com/max_pronko/status/737406159301292032

https://github.com/magento/magento2/issues/4840

**Comment by fooman** [comment]
An example of my own extensions - this 1 line change here
https://github.com/magento/magento2/commit/e81257235a89109012fdf5e2f038bf95272ad771#diff-f479129e7d10b50309fc7c8773e636cbL34
is responsible for breaking 4 of my M2 extensions.

![breakage-in-rc2](https://cloud.githubusercontent.com/assets/455508/15969285/ab4aabb8-2f2f-11e6-9314-5ca78400ab0f.png)

**Comment by antonkril** [comment]
Status update:
- Structure of UI components was added to definition of public stable api of a module. (not only @api annotated code is considered public api of module)
- We're still discussing your proposal & investigating options of making upgrades less painful while keeping promised public @api limited:
  - Current approach is to annotate much more code with @api
  - We're minimizing breaking changes of non-@api code. It's enforced by automatic tool. Every method/interface removal/modification is reviewed by architect
  - Working on tooling that will help identify changes in non-@api code that will break your extension
- Known breaking changes in 2.1 were [documented](http://devdocs.magento.com/guides/v2.1/release-notes/backward-incompatible-changes-2.1.html#changes-in-ui).

**Comment by veloraven** [comment]
I'm closing this issue as there was no response from author for a long time.
If this issue is still actual and you want create a new one GitHub issue please note that  we recently made some changes to the way we process GitHub submissions to more quickly identify and respond to core code issues. Feature Requests and Improvements should now be submitted to the new  [Magento 2 Feature Requests and Improvements forum](https://community.magento.com/t5/Magento-2-Feature-Requests-and/idb-p/feature-requests) (see details [here](https://community.magento.com/t5/News-Announcements/Improvements-to-GitHub-Management/m-p/44572#M96)).
